gazebo_ros/include/gazebo_ros/node.hpp

```diff
@@ -0,0 +1,152 @@
+   * \code{.xml}
+   * <!-- Optional configurations for a plugin's Node -->
+   * <ros>
+   *  <!-- Name of node, only use if plugin is one node -->
+   *  <node_name></node_name>
+   *  <!-- Namespace of the node -->
+   *  <namespace></namespace>
+   *  <!-- Command line arguments sent to Node's constructor for remappings -->
+   *  <arguments>
+   *     <argument>my_topic:=new_topic</argument>
+   *     <argument>__name:=super_cool_node</argument>
+   *  </arguments>
+   *  <!-- Initial ROS params set for node -->
+   *  <parameters>
+   *     <parameter name="max_velocity">55</parameter>
+   *     <parameter name="publish_odom">True</parameter>
+   *  </parameters>
+   * </ros>
+   * \endcode
+   * \param[in] node_name Name of node to create
+   * \param[in] _sdf An SDF element which either is a <ros> element or contains a <ros> element
+   * \return A shared pointer to a new #gazebo_ros::Node
+   */
+  static SharedPtr Create(const std::string & node_name, sdf::ElementPtr _sdf);
+
+  /// Create a #gazebo_ros::Node and add it to the global #gazebo_ros::Executor.
+  /**
+   * \note Must NOT called until #InitROS has been called
+   * \details Forwards arguments to the constructor for rclcpp::Node
+   * \param[in] args List of arguments to pass to <a href="http://docs.ros2.org/latest/api/rclcpp/classrclcpp_1_1_node.html">rclcpp::Node</a>
+   * \return A shared pointer to a new #gazebo_ros::Node
+   */
+  template<typename ... Args>
+  static SharedPtr Create(Args && ... args);
+
+  /// Initialize ROS with the command line arguments.
+  /**
+   * \note Must be called ONCE before any #gazebo_ros::Node instances are created. Subsequent calls are ignored.
+   * \param[in] argc Number of arguments
+   * \param[in] argv Vector of c-strings of length \a argc
+   */
+  static void InitROS(int argc, char ** argv);
+
+private:
+  /// Inherit constructor
+  using rclcpp::Node::Node;
+
+  /// Points to #static_executor_, so that when all #gazebo_ros::Node instances are destroyed, the
+  /// executor thread is too
+  std::shared_ptr<Executor> executor_;
+
+  /// True if #InitROS has been called and future calls will be ignored
+  static std::atomic_bool initialized_;
+
+  /// Locks #initialized_ and #executor_
+  static std::mutex lock_;
+
+  /// Points to an #gazebo_ros::Executor shared between all #gazebo_ros::Node instances
+  static std::weak_ptr<Executor> static_executor_;
+};
+
+template<typename ... Args>
+Node::SharedPtr Node::Create(Args && ... args)
+{
+  // Throw exception is Node is created before ROS is initialized
+  if (!initialized_) {
+    throw NotInitializedException();
+  }
+
+  // Contruct Node by forwarding arguments
+  Node::SharedPtr node = std::make_shared<Node>(std::forward<Args>(args) ...);
+
+  std::lock_guard<std::mutex> l(lock_);
+  // Store shared pointer to static executor in this object
+  node->executor_ = static_executor_.lock();
+  // If executor has not been contructed yet, do so now
+  if (!node->executor_) {
```

Comment:
If I'm understanding correctly, you may want to check `if(!node->static_executor_)` here instead.